### PR TITLE
PHP: Backport changes from core theme resolver.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## What?
Backport changes from https://github.com/WordPress/wordpress-develop/commit/8368eef8e00e3f9ab8a2bf2ca7093b07e05040b5 by @felixarntz

### CODE REVIEW

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey!

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 1]
```diff
@@ -34,11 +34,15 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
 		}
 
 		// When backporting to core, remove the instanceof Gutenberg class check, as it is only required for the Gutenberg plugin.
-		if ( null === static::$theme || ! static::$theme instanceof WP_Theme_JSON_Gutenberg ) {
+		if ( null === static::$theme || ! static::has_same_registered_blocks( 'theme' ) ) {
```
Comment (oandregal) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:37]: I don't know why there is the check for `static::$theme instanceof WP_Theme_JSON_Gutenberg`, though we should aim to understand that before removing it. What was the intention behind it? Is it safe to remove it?

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 2]
```diff
@@ -34,11 +34,15 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
...
+			$theme_json_file = static::get_file_path_from_theme( 'theme.json' );
 			$wp_theme        = wp_get_theme();
-			$theme_json_data = static::read_json_file( static::get_file_path_from_theme( 'theme.json' ) );
-			$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
-			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
```
Comment (oandregal) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:41]: This webfonts call is missing.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php (hunk 1 above)
Comment (spacedmonkey) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:37]: It is this way in core. This just a back port.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 3]
```diff
@@ -34,11 +34,15 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
...
+			if ( '' !== $theme_json_file ) {
+				$theme_json_data = static::read_json_file( $theme_json_file );
+				$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			} else {
+				$theme_json_data = array();
+			}
```
Comment (spacedmonkey) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:47]:
```suggestion
			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
```

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php (hunk 2 above)
Comment (spacedmonkey) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:41]: Fixed.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 4]
```diff
@@ -34,12 +34,16 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
...
-			$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			if ( '' !== $theme_json_file ) {
+				$theme_json_data = static::read_json_file( $theme_json_file );
+				$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			} else {
+				$theme_json_data = array();
+			}
 			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
-
 			/**
```
Comment (spacedmonkey) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php:48]:
```suggestion
			
			/**
```

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 5]
```diff
@@ -34,12 +34,17 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
...
-			$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			if ( '' !== $theme_json_file ) {
+				$theme_json_data = static::read_json_file( $theme_json_file );
+				$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			} else {
+				$theme_json_data = array();
+			}
 			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
-
+
```
Comment (spacedmonkey) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]:
```suggestion

```

**Comment by mirka** [comment]
Hi 👋 This may be a stupid question, but my wp-env is broken after this commit. Is there something I need to do to make it work?

For example, when I run `npm run wp-env start -- --update` on this commit, it fails with this error:

<details>
<summary>error</summary>

```
[07-Dec-2022 12:52:51 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method WP_Theme_JSON_Resolver_Gutenberg::has_same_registered_blocks() in /var/www/html/wp-content/plugins/gutenberg/lib/experimental/class-wp-theme-json-resolver-gutenberg.php:37
Stack trace:
#0 /var/www/html/wp-content/plugins/gutenberg/lib/compat/wordpress-6.2/class-wp-theme-json-resolver-6-2.php(161): WP_Theme_JSON_Resolver_Gutenberg::get_theme_data()
#1 /var/www/html/wp-content/plugins/gutenberg/lib/experimental/register-webfonts-from-theme-json.php(13): WP_Theme_JSON_Resolver_6_2::get_merged_data()
#2 /var/www/html/wp-includes/class-wp-hook.php(307): gutenberg_register_webfonts_from_theme_json('')
#3 /var/www/html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
#4 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#5 /var/www/html/wp-settings.php(598): do_action('init')
#6 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1336): require('/var/www/html/w...')
#7 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1254): WP_CLI\Runner->load_wordpress()
#8 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(28): WP_CLI\Runner->start()
#9 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(78): WP_CLI\Bootstrap\LaunchRunner->process(Object(WP_CLI\Bootstrap\BootstrapState))
#10 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(32): WP_CLI\bootstrap()
#11 phar:///usr/local/bin/wp/php/boot-phar.php(11): include('phar:///usr/loc...')
#12 /usr/local/bin/wp(4): include('phar:///usr/loc...')
#13 {main}
  thrown in /var/www/html/wp-content/plugins/gutenberg/lib/experimental/class-wp-theme-json-resolver-gutenberg.php on line 37
Fatal error: Uncaught Error: Call to undefined method WP_Theme_JSON_Resolver_Gutenberg::has_same_registered_blocks() in /var/www/html/wp-content/plugins/gutenberg/lib/experimental/class-wp-theme-json-resolver-gutenberg.php:37
Stack trace:
#0 /var/www/html/wp-content/plugins/gutenberg/lib/compat/wordpress-6.2/class-wp-theme-json-resolver-6-2.php(161): WP_Theme_JSON_Resolver_Gutenberg::get_theme_data()
#1 /var/www/html/wp-content/plugins/gutenberg/lib/experimental/register-webfonts-from-theme-json.php(13): WP_Theme_JSON_Resolver_6_2::get_merged_data()
#2 /var/www/html/wp-includes/class-wp-hook.php(307): gutenberg_register_webfonts_from_theme_json('')
#3 /var/www/html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
#4 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#5 /var/www/html/wp-settings.php(598): do_action('init')
#6 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1336): require('/var/www/html/w...')
#7 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1254): WP_CLI\Runner->load_wordpress()
#8 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(28): WP_CLI\Runner->start()
#9 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(78): WP_CLI\Bootstrap\LaunchRunner->process(Object(WP_CLI\Bootstrap\BootstrapState))
#10 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(32): WP_CLI\bootstrap()
#11 phar:///usr/local/bin/wp/php/boot-phar.php(11): include('phar:///usr/loc...')
#12 /usr/local/bin/wp(4): include('phar:///usr/loc...')
#13 {main}
  thrown in /var/www/html/wp-content/plugins/gutenberg/lib/experimental/class-wp-theme-json-resolver-gutenberg.php on line 37
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress. There has been a critical error on this website.
```
</details>

With the previous commit, it works fine.

**Comment by mirka** [comment]
Never mind, a complete wipe fixed it (`npm run wp-env destroy && npm run wp-env start`). Maybe it was just my env!

**Comment by spacedmonkey** [comment]
> Hi 👋 This may be a stupid question, but my wp-env is broken after this commit. Is there something I need to do to make it work?
> 
> For example, when I run `npm run wp-env start -- --update` on this commit, it fails with this error:
> 
> error
> With the previous commit, it works fine.

There is a fix here - https://github.com/WordPress/gutenberg/pull/46363 CC @mirka

**Comment by noahtallen** [comment]
> Never mind, a complete wipe fixed it (npm run wp-env destroy && npm run wp-env start). Maybe it was just my env!

Btw, I think this fixed it because it would have updated the WordPress source code, which introduces the missing properties that caused the crash! :) 

You can also run `npx wp-env start --update` to do that -- just another debug step that can be a bit faster than starting from scratch!

**Comment by mirka** [comment]
> Btw, I think this fixed it because it would have updated the WordPress source code, which introduces the missing properties that caused the crash! :)
> 
> You can also run `npx wp-env start --update` to do that -- just another debug step that can be a bit faster than starting from scratch!

I have to think there was something else involved because the `--update` flag wasn't fixing it. Given that nobody has commented here with the same problem, maybe there was something going on with my local files, I have no idea 🤷

**Comment by noahtallen** [comment]
I just had the exact same error myself trying to run `trunk` on WordPress 6.0 😅 I ended up switching to the last plugin release branch, because I specifically needed to test it on the older WP version 😬

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php (hunk 1 above)
Comment (oandregal) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php:37]: @spacedmonkey I've looked up git history and that check has been added in this PR https://github.com/WordPress/gutenberg/pull/42756 

As my understanding goes, without that check, there's an issue with some data not being recalculated in the plugin, so this PR has introduced a regression.

I am working on something that will fix the root issue (re: reorganizing theem.json code in the plugin), though, in the meantime, this is just such a small change that is worth adding it back until that lands.

**Pull request merged**